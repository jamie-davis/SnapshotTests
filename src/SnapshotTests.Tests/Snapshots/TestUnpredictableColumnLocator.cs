using System.Linq;
using NUnit.Framework;
using SnapshotTests.Snapshots;
using SnapshotTests.Tests.Snapshots.TestSupportUtils;
using TestConsole.OutputFormatting;
using TestConsoleLib;
using TestConsoleLib.Testing;

namespace SnapshotTests.Tests.Snapshots
{
    [TestFixture]
    public class TestUnpredictableColumnLocator
    {
        private SnapshotObjectMother _om;

        public TestUnpredictableColumnLocator()
        {
            _om = new SnapshotObjectMother();
        }

        [Test]
        public void UnpredictableColumnsAreExtracted()
        {
            //Arrange
            var tables = _om.Collection.TablesInDefinitionOrder.ToList();

            //Act
            var result = UnpredictableColumnLocator.Locate(tables);

            //Assert
            var output = new Output();
            var report = result.AsReport(rep => rep.Title("Extracted Unpredictable Columns")
                .RemoveBufferLimit()
                .AddColumn(t => t.Table.TableName)
                .AddChild(t => t.Columns, colRep => colRep.RemoveBufferLimit()
                    .AddColumn(c => c.Column.Name, cc => cc.Heading("Unpredictable Column"))
                    .AddChild(c => c.ReferencingColumns, refRep => refRep.Title("Referenced By")
                        .RemoveBufferLimit()
                        .AddColumn(r => r.SourceTable.TableName, cc => cc.Heading("Table"))
                        .AddColumn(rc => string.Join(", ", rc.References.Select(r => r.Name)), cc => cc.Heading("Columns")))));
            output.FormatTable(report);
            output.Report.Verify();

        }
    }
}