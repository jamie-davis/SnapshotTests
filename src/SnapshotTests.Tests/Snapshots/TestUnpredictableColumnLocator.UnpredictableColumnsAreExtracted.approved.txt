Extracted Unpredictable Columns                                                                
-------------------------------                                                                
Table Name                                                                                     
-----------------                                                                              
Table                                                                                          
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    Id                                                                                         
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table            Columns                                                               
        ---------------- --------                                                              
        ReferencingTable ParentId                                                              
                                                                                               
                                                                                               
Table Name                                                                                     
-----------------                                                                              
Table2                                                                                         
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    Id                                                                                         
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table             Columns                                                              
        ----------------- --------                                                             
        ReferencingTable2 ParentId                                                             
                                                                                               
                                                                                               
Table Name                                                                                     
-----------------                                                                              
ReferencingTable                                                                               
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    Id                                                                                         
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
                                                                                               
Table Name                                                                                     
-----------------                                                                              
ReferencingTable2                                                                              
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    Id                                                                                         
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
                                                                                               
Table Name                                                                                     
-----------------                                                                              
GuidKeyTable                                                                                   
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    GuidKeyId                                                                                  
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table        Columns                                                                   
        ------------ ---------------------------------                                         
        GuidRefTable GuidKeyParentId, GuidKeyParentId2                                         
                                                                                               
                                                                                               
Table Name                                                                                     
-----------------                                                                              
GuidRefTable                                                                                   
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    GuidKeyId                                                                                  
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
                                                                                               
Table Name                                                                                     
-----------------                                                                              
Unpredictables                                                                                 
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    GuidKeyId                                                                                  
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    Int                                                                                        
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    NullableInt                                                                                
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
    Unpredictable                                                                              
    Column                                                                                     
    -------------                                                                              
    TimeStamp                                                                                  
                                                                                               
        Referenced By                                                                          
        -------------                                                                          
        Table Columns                                                                          
        ----- -------                                                                          
                                                                                               
                                                                                               
                                                                                               